distributedcombigrid/src/sgpp/distributedcombigrid/manager/ProcessGroupWorker.cpp

```diff
@@ -570,6 +587,476 @@ void ProcessGroupWorker::combineUniform() {
+  }
+  assert( combiParametersSet_ );
+  int numGrids = combiParameters_.getNumGrids(); //we assume here that every task has the same number of grids
+
+  DimType dim = combiParameters_.getDim();
+  LevelVector lmin = combiParameters_.getLMin();
+  LevelVector lmax = combiParameters_.getLMax();
+  const std::vector<bool>& boundary = combiParameters_.getBoundary();
+
+  // the dsg can be smaller than lmax because the highest subspaces do not have
+  // to be exchanged
+  // todo: use a flag to switch on/off optimized combination
+
+  reduceSparseGridCoefficients(lmax,lmin,combiParameters_.getNumberOfCombinations(),currentCombi_,
+      combiParameters_.getLMinReductionVector(), combiParameters_.getLMaxReductionVector());
+  /*for (size_t i = 0; i < lmax.size(); ++i)
+        if (lmin[i] > 1)
+          lmin[i] -= 01;
+  for (size_t i = 0; i < lmax.size(); ++i)
+      lmax[i] = std::max(lmin[i],lmax[i] - 2);
+  */
+#ifdef DEBUG_OUTPUT
+  MASTER_EXCLUSIVE_SECTION{
+    std::cout << "lmin: "<< lmin << std::endl;
+    std::cout << "lmax: "<< lmax << std::endl;
+  }
+#endif
+
+  //delete old dsgs
+  combinedUniDSGVector_.clear();
+  // erzeug dsgs
+  combinedUniDSGVector_.resize(numGrids);
+  for (auto& uniDSG : combinedUniDSGVector_) {
+    uniDSG = std::unique_ptr<DistributedSparseGridUniform<CombiDataType>>(
+        new DistributedSparseGridUniform<CombiDataType>(dim, lmax, lmin, boundary,
+                                                        theMPISystem()->getLocalComm()));
+  }
+  // todo: move to init function to avoid reregistering
+  // register dsgs in all dfgs
+  for (Task* t : tasks_) {
+    for(int g=0; g<numGrids; g++){
+
+      DistributedFullGrid<CombiDataType>& dfg = t->getDistributedFullGrid(g);
+
+      dfg.registerUniformSG(*(combinedUniDSGVector_[g]));
+    }
+  }
+  Stats::stopEvent("combine init");
+  Stats::startEvent("combine hierarchize");
+
+  real localMax(0.0);
+  //std::vector<CombiDataType> beforeCombi;
+  for (Task* t : tasks_) {
+    t->fullgridVectorBeforeCombi = new std::vector<CombiDataType>[numGrids];
+    for(int g=0; g<numGrids; g++){
+
+      DistributedFullGrid<CombiDataType>& dfg = t->getDistributedFullGrid(g);
+      std::vector<CombiDataType> datavector(dfg.getElementVector());
+      t->fullgridVectorBeforeCombi[g] = datavector;
+      
+      // hierarchize dfg
+      DistributedHierarchization::hierarchize<CombiDataType>(
+          dfg, combiParameters_.getHierarchizationDims() );
+      
+
+      // lokales reduce auf sg ->
+      dfg.addToUniformSG( *combinedUniDSGVector_[g], combiParameters_.getCoeff( t->getID() ) );
+#ifdef DEBUG_OUTPUT
+      std::cout << "Combination: added task " << t->getID() << " with coefficient " << combiParameters_.getCoeff( t->getID() ) <<"\n";
+#endif
+    }
+  }
+  Stats::stopEvent("combine hierarchize");
+
+  Stats::startEvent("combine global reduce init");
+
+  Task::bufAsync = new std::vector<CombiDataType>[numGrids];
```

Comment:
memory leak?

Reply:
This could be the case. We could add a delete before. Or switch to smart pointers.

Reply:
or maybe make this a member of `DistributedFullGridUniform`? that is also connected to the next question...

Reply:
The buffer is deleted in line 854

Reply:
We could also put this as a member to the fullgrid directly. But I am not sure if this is improves things.